storage/remote/queue_manager.go

```diff
@@ -849,58 +974,77 @@ func (s *shards) sendSamples(ctx context.Context, samples []prompb.TimeSeries, b
 
 // sendSamples to the remote storage with backoff for recoverable errors.
 func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.TimeSeries, buf *[]byte) error {
-	backoff := s.qm.cfg.MinBackoff
-	req, highest, err := buildWriteRequest(samples, *buf)
+	req, highest, err := buildWriteRequest(samples, nil, *buf)
 	*buf = req
 	if err != nil {
 		// Failing to build the write request is non-recoverable, since it will
 		// only error if marshaling the proto to bytes fails.
 		return err
 	}
 
+	reportFailureFunc := func() {
+		s.qm.retriedSamplesTotal.Add(float64(len(samples)))
+	}
+	elapsed, err := sendWriteRequestWithBackoff(ctx, s.qm.cfg, s.qm.client, s.qm.logger, req, reportFailureFunc)
+	if elapsed > 0 {
+		s.qm.sentBatchDuration.Observe(elapsed)
+	}
+	if err != nil {
+		return err
+	}
+	s.qm.succeededSamplesTotal.Add(float64(len(samples)))
+	s.qm.bytesSent.Add(float64(len(*buf)))
+	s.qm.highestSentTimestampMetric.Set(float64(highest / 1000))
+	atomic.StoreInt64(&s.qm.lastSendTimestamp, time.Now().Unix())
+	return nil
+}
+
+func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, s StorageClient, l log.Logger, req []byte, reportFailureFunc func()) (elapsed float64, err error) {
+	backoff := cfg.MinBackoff
+
 	for {
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
+			return -1, ctx.Err()
 		default:
 		}
-		begin := time.Now()
-		err := s.qm.client.Store(ctx, req)
 
-		s.qm.sentBatchDuration.Observe(time.Since(begin).Seconds())
+		begin := time.Now()
+		err := s.Store(ctx, req)
+		elapsed = time.Since(begin).Seconds()
 
 		if err == nil {
-			s.qm.succeededSamplesTotal.Add(float64(len(samples)))
-			s.qm.bytesSent.Add(float64(len(req)))
-			s.qm.highestSentTimestampMetric.Set(float64(highest / 1000))
-			atomic.StoreInt64(&s.qm.lastSendTimestamp, time.Now().Unix())
-			return nil
+			return elapsed, nil
 		}
 
 		if _, ok := err.(recoverableError); !ok {
-			return err
+			return -1, err
 		}
-		s.qm.retriedSamplesTotal.Add(float64(len(samples)))
-		level.Debug(s.qm.logger).Log("msg", "failed to send batch, retrying", "err", err)
+
+		reportFailureFunc()
+		level.Debug(l).Log("msg", "failed to send batch, retrying", "err", err)
 
 		time.Sleep(time.Duration(backoff))
 		backoff = backoff * 2
-		if backoff > s.qm.cfg.MaxBackoff {
-			backoff = s.qm.cfg.MaxBackoff
+
+		if backoff > cfg.MaxBackoff {
+			backoff = cfg.MaxBackoff
 		}
 	}
 }
 
-func buildWriteRequest(samples []prompb.TimeSeries, buf []byte) ([]byte, int64, error) {
+func buildWriteRequest(samples []prompb.TimeSeries, metadata []prompb.MetricMetadata, buf []byte) ([]byte, int64, error) {
 	var highest int64
 	for _, ts := range samples {
 		// At the moment we only ever append a TimeSeries with a single sample in it.
 		if ts.Samples[0].Timestamp > highest {
 			highest = ts.Samples[0].Timestamp
 		}
 	}
+
 	req := &prompb.WriteRequest{
 		Timeseries: samples,
+		Metadata:   metadata,
```

Comment:
So actually metadata are NOT send together with normal samples remote write. That's might be surprising. (I just learned that ;p)

Reply:
Maybe picky nit but having different functions for meta and for samples might show that separation more explicitly?

Reply:
Initially, they were separate, however, it was suggested in a [previous comment](https://github.com/prometheus/prometheus/pull/6815#discussion_r386498861) to unify them - I have added a comment as we build the request to make it more explicit and also added a new comment on the proto to hint this.